src/qat/backend/analysis_passes.py

```diff
@@ -179,85 +177,196 @@ def extract_iter_bound(name: str, value: Union[List, np.ndarray]):
+            elif isinstance(inst, QuantumInstruction):
+                # TODO - recursive attribute lookup
+                read_vars = {
+                    k: v for k, v in inst.__dict__.items() if isinstance(v, Variable)
+                }
+                for var in read_vars.values():
+                    if not (
+                        var.name in result.symbol2scopes
+                        and [
+                            (s, e)
+                            for (s, e) in result.symbol2scopes[var.name]
+                            if s in stack
+                        ]
                     ):
-                        raise ValueError(f"Variable {name} already declared")
-                    result.iter_bounds[t].append(bound)
-            elif isinstance(inst, DeviceUpdate) and isinstance(inst.value, Variable):
-                if not (
-                    bound := next(
-                        (
-                            b
-                            for b in result.iter_bounds[inst.target]
-                            if b.name == inst.value.name
-                        ),
-                        None,
-                    )
-                ):
-                    raise ValueError(
-                        f"Variable {inst.value.name} referenced but no prior declaration found"
-                    )
-
-                result.var_binding[bound.name].append(inst)
-                if inst.attribute == "frequency":
-                    # TODO - guard against fixed_if
-                    concrete_bounds[inst.target].append(
-                        IterBound(
-                            name=inst.value.name,
-                            start=self.decompose_freq(bound.start, inst.target)[1],
-                            step=bound.step,
-                            end=self.decompose_freq(bound.end, inst.target)[1],
-                            count=bound.count,
+                        raise ValueError(
+                            f"Variable {var.name} referenced but no prior declaration found"
                         )
-                    )
-                else:
-                    raise NotImplementedError(
-                        f"Unsupported processing of attribute {inst.attribute}"
-                    )
 
-        for t, bounds in concrete_bounds.items():
-            groups = defaultdict(set)
-            for b in bounds:
-                groups[b.name].add(b)
+                if read_vars:
+                    result.reads[inst] = read_vars
+
+                if isinstance(inst, Acquire):
+                    result.writes[inst] = {
+                        inst.output_variable: Variable(inst.output_variable)
+                    }
+        if stack:
+            raise ValueError(f"Unbalanced scopes. Found orphans {stack}")
 
-            for name, grp in groups.items():
-                if len(grp) > 1:
+        res_mgr.add(result)
+
+
+class TILegalisationPass(AnalysisPass):
```

Comment:
Could we avoid acronyms here! Also is target-independent really that clear? It seems its doing SweepLegalisation no?